File: core/modules/worker_split.cc
// Copyright (c) 2014-2016, The Regents of the University of California.
// Copyright (c) 2016-2017, Nefeli Networks, Inc.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice, this
// list of conditions and the following disclaimer.
//
// * Redistributions in binary form must reproduce the above copyright notice,
// this list of conditions and the following disclaimer in the documentation
// and/or other materials provided with the distribution.
//
// * Neither the names of the copyright holders nor the names of their
// contributors may be used to endorse or promote products derived from this
// software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#include "worker_split.h"

const Commands WorkerSplit::cmds = {
    {"reset", "WorkerSplitArg", MODULE_CMD_FUNC(&WorkerSplit::CommandReset),
     Command::THREAD_UNSAFE}};

CommandResponse WorkerSplit::Init(const bess::pb::WorkerSplitArg &arg) {
  return CommandReset(arg);
}

CommandResponse WorkerSplit::CommandReset(const bess::pb::WorkerSplitArg &arg) {
  if (arg.worker_gates().empty()) {
    for (int i = 0; i < Worker::kMaxWorkers; i++) {
      gates_[i] = i;
    }
    return CommandSuccess();
  }

  for (size_t i = 0; i < Worker::kMaxWorkers; i++) {
    gates_[i] = -1;
  }

  for (auto it : arg.worker_gates()) {
    gate_idx_t ogate = it.second;
    if (ogate >= MAX_GATES) {
      return CommandFailure(EINVAL, "output gate must be less than %" PRIu16,
                            MAX_GATES);
    }
    gates_[it.first] = ogate;
  }

  return CommandSuccess();
}

void WorkerSplit::ProcessBatch(Context *ctx, bess::PacketBatch *batch) {
  int gate = gates_[ctx->wid];
  if (gate >= 0) {
    RunChooseModule(ctx, gate, batch);
  } else {
    bess::Packet::Free(batch);
  }
}

void WorkerSplit::AddActiveWorker(int wid, const Task *t) {
  if (!HaveVisitedWorker(t)) {  // Have not already accounted for worker.
    active_workers_[wid] = true;
    visited_tasks_.push_back(t);
    // Only propagate workers downstream on ogate mapped to `wid`
    int g = gates_[wid];
    bess::OGate *ogate = (g < 0) ? nullptr : ogates()[g];
    if (ogate) {
      auto next = static_cast<Module *>(ogate->next());
      next->AddActiveWorker(wid, t);
    }
  }
}

ADD_MODULE(WorkerSplit, "ws",
           "send packets to output gate X, the id of current worker")

File: bessctl/module_tests/worker_split.py
# Copyright (c) 2016-2017, Nefeli Networks, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# * Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.
#
# * Neither the names of the copyright holders nor the names of their
# contributors may be used to endorse or promote products derived from this
# software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import multiprocessing as mp
from test_utils import *


class BessWorkerSplitTest(BessModuleTestCase):

    def test_worker_split_default(self):
        NUM_WORKERS = 2

        for wid in range(NUM_WORKERS):
            for i in range(NUM_WORKERS):
                bess.add_worker(wid=i, core=i)

            src = Source()
            ws = WorkerSplit()
            src -> ws

            for i in range(NUM_WORKERS):
                ws:i -> Sink()

            src.attach_task(wid=wid)

            bess.resume_all()
            time.sleep(1)
            bess.pause_all()

            # packets should flow onto only one output gate...
            ogates = bess.get_module_info(ws.name).ogates
            for ogate in ogates:
                if ogate.ogate == wid:
                    self.assertGreater(ogate.pkts, 0)
                else:
                    self.assertEquals(ogate.pkts, 0)

            bess.reset_all()

    def test_worker_split_fancy(self):
        NUM_WORKERS = mp.cpu_count()

        gates = dict()
        for i in range(NUM_WORKERS):
            gates[i] = i & 1

        for i in range(NUM_WORKERS):
            bess.add_worker(wid=i, core=i)
            bess.add_tc('rl_{}'.format(i), policy='rate_limit', wid=i,
                        resource='count', limit={'count': 1})
        bess.pause_all()

        ws::WorkerSplit(worker_gates=gates)
        ws:0 -> sink0::Sink()
        ws:1 -> sink1::Sink()

        srcs = [Source() for _ in range(NUM_WORKERS)]
        for i in range(NUM_WORKERS):
            srcs[i].attach_task(parent='rl_{}'.format(i))
            srcs[i] -> ws

        bess.resume_all()
        time.sleep(3)
        bess.pause_all()

        odd_pkts = 0
        even_pkts = 0
        for i in range(NUM_WORKERS):
            pkts = bess.get_module_info(srcs[i].name).ogates[0].pkts
            if i % 2 == 0:
                even_pkts += pkts
            else:
                odd_pkts += pkts

        ws_ogates = bess.get_module_info(ws.name).ogates
        for ogate in ws_ogates:
            if ogate.ogate == 0:
                self.assertEquals(ogate.pkts, odd_pkts)
            elif ogate.ogate == 1:
                self.assertEquals(ogate.pkts, odd_pkts)

        bess.reset_all()

suite = unittest.TestLoader().loadTestsFromTestCase(BessWorkerSplitTest)
results = unittest.TextTestRunner(verbosity=2).run(suite)

if results.failures or results.errors:
    sys.exit(1)
